### PR TITLE
Remote Lock Interface

### PR DESCRIPTION
**Pull request by vaibhawvipul**
resolves - https://github.com/resonatehq/resonate-sdk-ts/issues/25

- implementation of lock interface in remote mode

### CODE REVIEW

**Comment by guergabo** [comment]
hi @vaibhawvipul, this pr will alter slightly the payload for acquire and heartbeat apis: https://github.com/resonatehq/resonate/pull/216. 

I'd recommend just looking at the openapi file which is up to date in the PR. 

If you have any questions lmk.

File: lib/core/stores/remote.ts [hunk 1]
```diff
@@ -12,22 +12,22 @@ import { IStore, IPromiseStore, IScheduleStore } from "../store";
...
   public promises: RemotePromiseStore;
   public schedules: RemoteScheduleStore;
-  public locks: LocalLockStore;
+  public locks: RemoteLockStore;
 
   constructor(url: string, logger: ILogger, encoder: IEncoder<string, string> = new Base64Encoder()) {
     this.promises = new RemotePromiseStore(url, logger, encoder);
     this.schedules = new RemoteScheduleStore(url, logger, encoder);
 
     // temp
```
Comment (dfarr) [inline lib/core/stores/remote.ts:29]: Mind getting rid of this comment?

File: lib/core/stores/remote.ts [hunk 2]
```diff
@@ -496,3 +496,110 @@ function isSearchSchedulesResult(obj: any): obj is { cursor: string; schedules:
     obj.schedules.every(isSchedule)
   );
 }
+
+export class RemoteLockStore implements ILockStore {
+  private url: string;
+  // simple lock counter - tryAcquire increments, release decrements
+  // counter > 0 we call heartbeat api in a loop, setInterval
+  private lockCounter: number = 0;
+  private heartbeatInterval: NodeJS.Timeout | number | null = null;
```
Comment (dfarr) [inline lib/core/stores/remote.ts:505]:
```suggestion
  private heartbeatInterval: number | null = null;
```

Let's just coerce this value to a number, you can do this by doing `this. heartbeatInterval  = +setInterval()`

File: lib/core/stores/remote.ts [hunk 3]
```diff
@@ -496,3 +494,97 @@ function isSearchSchedulesResult(obj: any): obj is { cursor: string; schedules:
...
+      if (locksAffected === 0) {
+        this.stopHeartbeat();
+      }
+    }, 5000); // desired heartbeat interval (in ms)
+  }
+
+  private stopHeartbeat(): void {
+    if (this.heartbeatInterval) {
+      clearInterval(this.heartbeatInterval);
+      this.heartbeatInterval = 0;
```
Comment (dfarr) [inline lib/core/stores/remote.ts:567]:
```suggestion
      this.heartbeatInterval = null;
```

**Review by dfarr — APPROVED** [review]

**Pull request merged**